### PR TITLE
Danfuzz cat

### PR DESCRIPTION
**Pull request by danfuzz**
Addition of the Cat class. See docs for details.

/cc @mikefleming 

R=mike

### CODE REVIEW

**Comment by danfuzz** [comment]
Note: Two of the commits (the "test" and "merge" commits) seem to reflect the fact that I was working on this code at the same time as the previous PR. I believe they'll turn into null-diff merges as this goes into master.

**Comment by danfuzz** [comment]
Hey, I still want a review, but in the name of forward progress (particularly, so I can use regular npm semvers in dependents), I'm going to go ahead and merge this for now.

Thanks for your indulgence!

File: lib/cat.js [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+ * Construct a Cat state object.
+ */
+function State(emitter, streams, paused) {
+    this.emitter = emitter;
+    this.streams = [];
+    this.paused = (paused === undefined) ? true : !!paused;
+    this.readable = true;
+
+    this.onCloseOrEnd = this.onCloseOrEnd.bind(this);
+    this.onData = this.onData.bind(this);
```
Comment (mikefleming) [inline lib/cat.js:38]: am confused: why are these necessary?

File: lib/cat.js [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        this.streams.push(blip);
+    }
+
+    if (!this.paused) {
+        this.streams[0].resume();
+    }
+}
+
+State.prototype.destroy = function destroy() {
+    this.emitter = undefined;
```
Comment (mikefleming) [inline lib/cat.js:80]: I'd actually assume this would destroy the other streams

Comment (danfuzz) [inline lib/cat.js:80]: I will clarify this in the docs.

The design of all of the classes in this module are as layers on top of other streams, where the other stream is listened to but not otherwise modified.

I've considered making pass-through be an option specifically for the flow control methods `pause()` and `resume()`; and `destroy()` may want to be optionally passed-through as well. However, I decided not to implement that until / unless I find an actual use case.

File: lib/cat.js (hunk 1 above)
Comment (danfuzz) [inline lib/cat.js:38]: I assume you're wondering why we need to `bind()` these methods.

The deal is that the methods get added as listeners to the argument streams (well, to a set of `Valve` instances on top of them). If they weren't bound, then the event callbacks would get called with an unpredictable `this`.

I'll add some comments (here and in the other equivalent spots).

**Comment by danfuzz** [comment]
Thanks for the comments. I'm opening a new PR to address them.
